src/libsystemd/sd-event/event-util.c

```diff
@@ -177,3 +177,61 @@ dual_timestamp* event_dual_timestamp_now(sd_event *e, dual_timestamp *ts) {
         assert_se(sd_event_now(e, CLOCK_MONOTONIC, &ts->monotonic) >= 0);
         return ts;
 }
+
+void event_source_unref_many(sd_event_source **array, size_t n) {
+        FOREACH_ARRAY(v, array, n)
+                sd_event_source_unref(*v);
+
+        free(array);
+}
+
+static int event_forward_signal_callback(sd_event_source *s, const struct signalfd_siginfo *ssi, void *userdata) {
+        sd_event_source *child = ASSERT_PTR(userdata);
+
+        assert(ssi);
+
+        siginfo_t si = {
+                .si_signo = ssi->ssi_signo,
+                /* This has to be SI_QUEUE as SI_USER and SI_KERNEL are not allowed to be used. */
+                .si_code = SI_QUEUE,
+                .si_errno = ssi->ssi_errno,
+        };
+
+        si.si_pid = ssi->ssi_pid;
+        si.si_uid = ssi->ssi_uid;
+        si.si_int = ssi->ssi_int;
+        si.si_ptr = UINT64_TO_PTR(ssi->ssi_ptr);
+
+        return sd_event_source_send_child_signal(child, ssi->ssi_signo, &si, /* flags = */ 0);
+}
+
+int event_forward_signals(sd_event *e, sd_event_source *child, sd_event_source ***ret_sources, size_t *ret_n_sources) {
```

Comment:
If you remove SIGHUP I think this should be moved to ptyfwd.c btw. Because it is not really that generic.

Reply:
I don't think it makes sense there either tbh, we might want this for more than just ptyfwd in the future.

I think even-util.c is fine for now. If we start using this in more places that need different signals, we can make the list of signals an input argument instead of hard coded.

Reply:
I mean ptyfwd.c, not -tool.

And I prefer to always keep stuff private unless it's certainly generic enough...

Reply:
Made the function more generic and moved the list of signals to ptyfwd.c